MediaBrowser.Controller/SyncPlay/PlaybackRequest/BufferGroupRequest.cs

```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Threading;
+using MediaBrowser.Model.SyncPlay;
+using MediaBrowser.Controller.Session;
+
+namespace MediaBrowser.Controller.SyncPlay
+{
+    /// <summary>
+    /// Class BufferingGroupRequest.
+    /// </summary>
+    public class BufferGroupRequest : IPlaybackGroupRequest
+    {
+        /// <summary>
+        /// Gets or sets when the request has been made by the client.
+        /// </summary>
+        /// <value>The date of the request.</value>
+        public DateTime When { get; set; }
+
+        /// <summary>
+        /// Gets or sets the position ticks.
+        /// </summary>
+        /// <value>The position ticks.</value>
+        public long PositionTicks { get; set; }
+
+        /// <summary>
+        /// Gets or sets the client playback status.
+        /// </summary>
+        /// <value>The client playback status.</value>
+        public bool IsPlaying { get; set; }
+
+        /// <summary>
+        /// Gets or sets the playlist item id of the playing item.
+        /// </summary>
+        /// <value>The playlist item id.</value>
+        public string PlaylistItemId { get; set; }
+
+        /// <inheritdoc />
+        public PlaybackRequestType GetRequestType()
+        {
+            return PlaybackRequestType.Buffer;
+        }
+
+        /// <inheritdoc />
+        public void Apply(ISyncPlayStateContext context, ISyncPlayState state, SessionInfo session, CancellationToken cancellationToken)
```

Comment:
From the look of the code - this method is identical across all GroupRequests.

using a base class, or interface extension would negate code repetition? Alternatively, could the code in the function be moved up one level?

Reply:
What I did here is try to apply the [Visitor pattern design](https://en.wikipedia.org/wiki/Visitor_pattern) where each **GroupState** is a **Visitor** and each **GroupRequest** is an **Element**. It simplifies the code a lot (by avoiding switch cases) at the cost of having to duplicate this method in every **Element**. It also decouples classes that should not be tied.

This design relies on the type of the class to switch between the right method to call. Creating an abstract class would make the switching always select the base class and not differentiate between the different **GroupRequest**s.

Also, it makes the code cleaner to read IMHO :smile:.